llvm/lib/Target/RISCV/RISCVExpandCoreVHwlpPseudoInsts.cpp

```diff
@@ -0,0 +1,209 @@
+        MachineFunctionPass::getAnalysisUsage(AU);
+      }
+  };
+}
+
+char RISCVExpandCoreVHwlpPseudo::ID = 0;
+
+bool RISCVExpandCoreVHwlpPseudo::runOnMachineFunction(MachineFunction &MF) {
+  if (!MF.getSubtarget<RISCVSubtarget>().hasExtXCoreVHwlp()) {
+    return false;
+  }
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  bool Changed = false;
+  for (auto &ML : MLI) {
+    Changed |= ProcessLoop(ML, MF);
+  }
+  return Changed;
+}
+
+bool RISCVExpandCoreVHwlpPseudo::ProcessLoop(MachineLoop *ML,
+                                             MachineFunction &MF) {
+  bool Changed = false;
+  for (auto &Inner : *ML) {
+    Changed |= ProcessLoop(Inner, MF);
+  }
+
+  MachineBasicBlock *Preheader = ML->getLoopPreheader(),
+                    *Latch = ML->getLoopLatch();
+
+  if (!Preheader || !Latch) {
+    return Changed;
+  }
+
+  MachineInstr *HwlpSetup = nullptr, *HwlpBranch = nullptr;
+  for (auto &MI : *Preheader) {
+    if (MI.getOpcode() == RISCV::HwlpSetup ||
+        MI.getOpcode() == RISCV::HwlpSetupImm) {
+      HwlpSetup = &MI;
+      break;
+    }
+  }
+  for (auto &MI : Latch->terminators()) {
+    if (MI.getOpcode() == RISCV::HwlpBranch) {
+      HwlpBranch = &MI;
+    }
+  }
+  if (!HwlpSetup || !HwlpBranch) {
+    return Changed;
+  }
+
+  const TargetInstrInfo *TII = MF.getSubtarget().getInstrInfo();
+
+  MachineBasicBlock *LoopHeader = ML->getHeader();
+  LoopHeader->setLabelMustBeEmitted();
+
+  // Iterate through the instructions to get the last instruction and the offset
+  // for the loop start and end.
+  int Offset = 0, StartOffset = 0, EndOffset = 0, InnerHwlpEndOffset = 0;
+  MCSymbol *InnerEndSymbol = nullptr;
+  MachineInstr *MI = HwlpSetup,
+               *LastInstr = nullptr;
+  while (MI) {
+    if (!StartOffset && MI->getParent() == LoopHeader) {
+      StartOffset = Offset;
+    }
+    if (MI == HwlpBranch) {
+      // Add 'nop' if the loop body is too short or the loops end is too close
+      // to the end of an inner hardware loop.
+      if (EndOffset - StartOffset < 8 || EndOffset - InnerHwlpEndOffset < 8) {
+        MI = BuildMI(*Latch, MI, MI->getDebugLoc(), TII->get(RISCV::ADDI),
+                     RISCV::X0).addReg(RISCV::X0).addImm(0);
+      } else {
+        break;
+      }
+    }
+    if (!MI->isDebugInstr()) {
+      LastInstr = MI;
+      EndOffset = Offset;
+    }
+    // Find the end of an inner hardware loop
+    switch (MI->getOpcode()) {
+      case RISCV::CV_SETUP:
+      case RISCV::CV_SETUPI:
+        InnerEndSymbol = MI->getOperand(2).getMCSymbol();
+        break;
+      case RISCV::CV_ENDI:
+        InnerEndSymbol = MI->getOperand(1).getMCSymbol();
+        break;
+      default:
+        break;
+    }
+    if (InnerEndSymbol && InnerEndSymbol == MI->getPreInstrSymbol()) {
+      InnerHwlpEndOffset = Offset;
+    }
+    unsigned Size = TII->getInstSizeInBytes(*MI);
+    // Instruction compression is disabled here
+    if (Size == 2){
```

Comment:
Isn't this just an error? How can a instruction of size 2 exist in a HWLP?

Reply:
`getInstSizeInBytes` doesn't know that instruction compression is not allowed here and returns 2 on compressible instructions.